fullscreen.bs

```diff
@@ -392,23 +400,38 @@ could be an open <{dialog}> element.
  <a>active document</a>.
  <!-- cross-process -->
 
+ <li><p>Let <var>pendingDoc</var> be <var>doc</var>.
+
+ <li><p>Let <var>resize</var> be false.
+
  <li><p>If <var>topLevelDoc</var> is in <var>docs</var>, and it is a
- <a>simple fullscreen document</a>, then set <var>doc</var> to <var>topLevelDoc</var> and
+ <a>simple fullscreen document</a>, then set <var>pendingDoc</var> to <var>topLevelDoc</var> and
  <var>resize</var> to true.
 
+ <li><p>Let <var>pending</var> be <var>pendingDoc</var>'s <a>fullscreen element</a>.
+ <!-- cross-process -->
+
  <li><p>Return <var>promise</var>, and run the remaining steps <a>in parallel</a>.
 
- <li><p>If <var>resize</var> is true, resize <var>doc</var>'s viewport to its "normal" dimensions.
+ <li><p>If <var>resize</var> is true, resize <var>pendingDoc</var>'s viewport to its "normal"
+ dimensions.
 
- <li><p>If <var>doc</var>'s <a>fullscreen element</a> is null, then resolve <var>promise</var> with
- undefined and terminate these steps.
+ <li>
+  <p>If <var>pendingDoc</var>'s <a>fullscreen element</a> is not <var>pending</var>:
+
+  <ol>
+   <li><p><a for=set>Append</a> (<code>fullscreenchange</code>, <var>pending</var>) to
+   <var>pendingDoc</var>'s <a>list of pending fullscreen events</a>.
+
+   <li><p>Resolve <var>promise</var> with undefined and terminate these steps.
+  </ol>
 
  <li><p>Let <var>exitDocs</var> be the result of
  <a lt="collect documents to unfullscreen">collecting documents to unfullscreen</a> given
- <var>doc</var>.
+ <var>pendingDoc</var>.
```

Comment:
Because `pending` is never null, any case where the fullscreen element is now null (or indeed different at all) is taken care or, and we "terminate these steps" above. The assert in "collect documents to unfullscreen" is in place in my changes to match these changes and isn't hit for this reason.

But why does it make sense to only fire an event for this document and not the ancestors in this case, one wonders? Because if the `resize` is true, then `pendingDoc` is the top-level document. And if we hit the above condition for that case, then it was an iframe that was removed and all the descendant documents have already been disposed of. At least I've been unable to come up with a case that's going to be broken, although that doesn't mean there isn't.  https://github.com/whatwg/fullscreen/issues/63 is a potential concern, but there I really think the fix should be that we make those kinds of combinations/races impossible, since I don't think there will always be a "reasonable" thing to do when we end up in those situations.

Reply:
Hmmm... It seems I somehow overlooked "terminate these steps" there...

Reply:
If we "terminate these steps" when `pending` is different than `pendingDoc`'s fullscreen element, there would be a weird case:

If we have top-level document A being: `<div id="outer"><iframe></iframe></div>`, then the `iframe` contains document B. Now we fullscreen `#outer` in A, then fullscreen `document.body` in B. When we remove `document.body` from B, it seems that the `iframe` in A would keep being in fullscreen.

However, if we don't fullscreen `#outer` at the beginning, removing `document.body` from B would trigger a full unfullscreen, and the event would be solely dispatched to A rather than B.

Reply:
OK, so this is a case where A will have `#outer` and `iframe` in top layer with the "fullscreen flag" set, but `iframe` doesn't have the the "iframe fullscreen flag" set. And B has just its `document.body` in top layer as the fullscreen element. Calling `document.exitFullscreen()` in B would normally leave us with with A having just `#outer` as the fullscreen element, i.e. we popped the `iframe` from top layer. But, with what I've proposed, we'd "terminate these steps" and document A would be left with two elements in top layer.

One possible stance is "oh well, removing the fullscreen element does weird things, just like exiting fullscreen twice can". But I'll take a look at your suggestions to tweak "collect documents to unfullscreen", to see if I can make that behave exactly as if the fullscreen element hadn't changed yet, kind of like how it was when the fullscreen element stack was separate.

Reply:
I think the second case is also an issue, that B doesn't get `fullscreenchange` and only A gets that.

I'm wondering whether it is easier to tweak this approach, or to somehow fully exit fullscreen... To fully exit fullscreen for fullscreen element removal, we can just queue an event on all descendant documents, and unfullscreen all of them. That may be easier than reusing exit fullscreen for element removal.

Reply:
> However, if we don't fullscreen `#outer` at the beginning, removing `document.body` from B would trigger a full unfullscreen, and the event would be solely dispatched to A rather than B.

I guess this is the second case, I didn't notice it was a separate issue before. I think what would happen per the changes now in this PR is that *resize* would be true, but after we've resized B no longer has a fullscreen element and *pendingDoc* is A and *pending* is the iframe in A, so no event fired in B, that's right.

To fully exit fullscreen couldn't fix this case I'm afraid, some bookkeeping would still be needed to remember that B had a fullscreen element and should get a `fullscreenchange` event after the resize.

It is quite tempting to suggest keeping track of the fullscreen element separately from the top layer, or even to revive the old fullscreen element stack. But I don't think that's a good idea, it will open up a new set of corner cases where the two things are out of sync.

There is one approach I haven't really explored yet, something analogous to https://github.com/whatwg/fullscreen/pull/64 but for exiting fullscreen. If we were to decide up front how to manipulate the top layer of each document, and in the async part are able to somehow handle the differences from what we expected the top layers to look like, then at least we ought not fire too *few* fullscreenchange events, although we might still be able to find cases where we'd fire redundant ones.

WDYT, worth trying?

Or should we look for something non-clever that often does the right thing and is known to be broken if you do something too strange?

Reply:
> To fully exit fullscreen couldn't fix this case I'm afraid, some bookkeeping would still be needed to remember that B had a fullscreen element and should get a fullscreenchange event after the resize.

Not necessarily?

I think we can change `fully exit fullscreen` to this:
1. Let topLevelDoc be document’s top-level browsing context’s active document.
2. If topLevelDoc ’s fullscreen element is null, terminate these steps. 
3. Let docs be an ordered set consisting of topLevelDoc’s descendant browsing contexts' active documents whose fullscreen element is non-null, in tree order.
4. For each doc in docs:
    * Append (fullscreenchange, doc's fullscreen element) to doc’s list of pending fullscreen events.
    * Unfullscreen doc.
5. Run the following step in parallel.
6. Resize topLevelDoc’s viewport to its "normal" dimensions.

Then we just need to call `fully exit fullscreen` in the removing steps when fullscreen element is found in it.

Since this `fully exit fullscreen` doesn't involve any asynchronous operation on top layer (either read or write), it can be safely used as the last resort cleanup for everything now. And this procedure should be more robust that it doesn't need much change however fullscreen works. (e.g. with or without hierarchy restriction)

> If we were to decide up front how to manipulate the top layer of each document, and in the async part are able to somehow handle the differences from what we expected the top layers to look like, 

It's not clear to me how it would look like... Everything I can think based on this feels overcomplicated...

> Or should we look for something non-clever that often does the right thing and is known to be broken if you do something too strange?

I don't think we should leave an active document being in a broken state... Edge cases can happen unintentionally (e.g. in some race condition), and we should handle that gracefully as much as possible...

Reply:
@dtapuska You mean the `fully exit fullscreen` algorithm above? No, it doesn't change when the event gets fired. The event is still queued in the `pending fullscreen events`, and fired at roughly the rAF time.

I don't quite understand your second question. Why do we want to check whether fullscreen element is connected when we are exiting fullscreen? Am I missing something?

Reply:
@upsuper I'm referring to the fact that you queue the `fullscreenchange` event before the viewport is resized asynchronously. The end page will see the `fullscreenchange` before the `resize` event. Whereas in Chrome at least the ordering is the `resize` event first then the `fullscreenchange` event (both sent at rAF time). But with your proposal you'll likely get
`fullscreenchange` in one rAF and a `resize` in another rAF.

So the second question is that you had proposed to change how when a node is removed from the document that you should dispatch unfullscreen the node, queue fullscreenchange events and resize the viewport. I'm indicating we should unfullscreen the node, resize the viewport then queue the fullscreenchange events. And to do this it would just be a twist on the `Exit fullscreen` algorithm so see if the fullscreen element (which is non-null) is connected. Again I'm new to this so I'm happy to take an answer that it doesn't make sense.

Reply:
Ok I tried to implemented @upsuper's suggestion and it appears to work however. I do see some difference in how the document cleanup steps run between Firefox and Chrome. When an iframe navigates to a new location Firefox exits fullscreen (matching what would the new fully exit fullscreen text) but Chrome doesn't. It keeps the fullscreen mode for the navigation to the new document. Here is a example http://output.jsbin.com/lonoyeg

Reply:
Edge seems to match this as well. So I think that the proposed text is probably superior.  @foolip are we ok to change how Chrome behaves here?

Reply:
I put the proposal here: #126

Reply:
Hmm, okay, that makes sense. So we should probably move the appending after step 5 or 6 in my proposal, so that fullscreen events would come after resize event.

Reply:
> Would it make sense to add a step in the exit fullscreen that indicated if the fullscreen element is not connected then Unfullscreen it immediately (say after step 4)?

@dtapuska, was that assuming some of the changes in this PR, i.e., that within the "removing steps", we first call "exit fullscreen" (or "fully exit fullscreen") and then remove the element from top layer before any of the async bits, including resizing, happen? I'm not sure which step 4 you mean, but everything needs to be done or decided in the synchronous bits, as `document.fullscreenElement` will change by the end of the "removing steps".

(Probably this was clear, just restating to make sure.)

Reply:
> I do see some difference in how the document cleanup steps run between Firefox and Chrome. When an iframe navigates to a new location Firefox exits fullscreen (matching what would the new fully exit fullscreen text) but Chrome doesn't.

@dtapuska I wonder if this comes down to the difference between "fully exit fullscreen" exiting fully for just the document it was invoked with, or the top-level document? The spec now has the first model, and https://github.com/whatwg/fullscreen/pull/126 changes it to the second.

Reply:
> Edge seems to match this as well

@dtapuska in other words, when navigating any document which is in fullscreen, Edge fully exits fullscreen?